integration/integration_test.go

```diff
@@ -0,0 +1,263 @@
+	leader, err := cl.Leader()
+	require.NoError(t, err)
+	require.NoError(t, cl.SetNodeRole(leader.node.NodeID(), api.NodeRoleWorker))
+	// agents 2, managers 2
+	numWorker++
+	numManager--
+	pollClusterReady(t, cl, numWorker, numManager)
+
+	// remove node
+	var worker *TestNode
+	for _, n := range cl.nodes {
+		if !n.config.IsManager && n.node.NodeID() != leader.node.NodeID() {
+			worker = n
+			break
+		}
+	}
+	require.NoError(t, cl.RemoveNode(worker.node.NodeID()))
+	// agents 1, managers 2
+	numWorker--
+	// long wait for heartbeat expiration
+	pollClusterReady(t, cl, numWorker, numManager)
+
+	// promote old leader back
+	require.NoError(t, cl.SetNodeRole(leader.node.NodeID(), api.NodeRoleManager))
+	numWorker--
+	numManager++
+	// agents 0, managers 3
+	pollClusterReady(t, cl, numWorker, numManager)
+}
+
+func TestDemotePromote(t *testing.T) {
+	numWorker, numManager := 1, 3
+	cl := newCluster(t, numWorker, numManager)
+	defer func() {
+		require.NoError(t, cl.Stop())
+	}()
+
+	leader, err := cl.Leader()
+	require.NoError(t, err)
+	var manager *TestNode
+	for _, n := range cl.nodes {
+		if n.config.IsManager && n.node.NodeID() != leader.node.NodeID() {
+			manager = n
+			break
+		}
+	}
+	require.NoError(t, cl.SetNodeRole(manager.node.NodeID(), api.NodeRoleWorker))
+	// agents 2, managers 2
+	numWorker++
+	numManager--
+	pollClusterReady(t, cl, numWorker, numManager)
+
+	// promote same node
+	require.NoError(t, cl.SetNodeRole(manager.node.NodeID(), api.NodeRoleManager))
+	// agents 1, managers 3
+	numWorker--
+	numManager++
+	pollClusterReady(t, cl, numWorker, numManager)
+}
+
+func TestPromoteDemote(t *testing.T) {
+	numWorker, numManager := 1, 3
+	cl := newCluster(t, numWorker, numManager)
+	defer func() {
+		require.NoError(t, cl.Stop())
+	}()
+
+	var worker *TestNode
+	for _, n := range cl.nodes {
+		if !n.config.IsManager {
+			worker = n
+			break
+		}
+	}
+	require.NoError(t, cl.SetNodeRole(worker.node.NodeID(), api.NodeRoleManager))
+	// agents 0, managers 4
+	numWorker--
+	numManager++
+	pollClusterReady(t, cl, numWorker, numManager)
+
+	// demote same node
+	require.NoError(t, cl.SetNodeRole(worker.node.NodeID(), api.NodeRoleWorker))
+	// agents 1, managers 3
+	numWorker++
+	numManager--
+	pollClusterReady(t, cl, numWorker, numManager)
+}
+
+func TestDemotePromoteLeader(t *testing.T) {
+	numWorker, numManager := 1, 3
+	cl := newCluster(t, numWorker, numManager)
+	defer func() {
+		require.NoError(t, cl.Stop())
+	}()
+
+	leader, err := cl.Leader()
+	require.NoError(t, err)
+	require.NoError(t, cl.SetNodeRole(leader.node.NodeID(), api.NodeRoleWorker))
+	// agents 2, managers 2
+	numWorker++
+	numManager--
+	pollClusterReady(t, cl, numWorker, numManager)
+
+	//promote former leader back
+	require.NoError(t, cl.SetNodeRole(leader.node.NodeID(), api.NodeRoleManager))
+	// agents 1, managers 3
+	numWorker--
+	numManager++
+	pollClusterReady(t, cl, numWorker, numManager)
+}
+
+func TestDemoteToSingleManager(t *testing.T) {
```

Comment:
See #1181 and #1178 we should probably add a `// TODO: improve test to demote the leader in case of 2 remaining managers`